enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,365 @@
+`ControlPlaneMachineSet` if scaling the nodes is determined to be required.
+
+The new configuration will specify if scaling up and down is enabled and allow
+configuring when to trigger scaling. It will also allow configuring which
+machine types are to be used for scaling, putting an upper and lower limit on
+possible automatic scaling.
+
+### Workflow Description
+
+1. The OpenShift adminstrator creates a valid
+   `control-plane-machine-set-autoscaling` `CR` in the `openshift-machine-api`
+   namespace (or the respective namespace `control-plane-machineset-operator` is
+   running in), to configure the automatic vertical scaling.
+
+### API Extensions
+
+This proposal requires a new custom resource to configure when and how
+automatically scale the control plane by modifying the control plane machinset.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The implementation for this feature, requires a data source to gather current
+CPU and memory usage for nodes, as well as available resources on the nodes.
+
+To gather available resource information the default
+[Node](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#node-v1-core)
+API endpoint can be used.
+
+Using the [horizontal pod
+autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/)
+as a reference, they are using the [metrics
+API](https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/)
+to gather performance statistics for nodes. 
+
+Using the metrics API has the least requirements as it does not expect a running
+Prometheus monitoring stack. Using Prometheus generated data could make the
+implementation easier with respect to data aggregation.
+
+Data aggregation will be required using the metrics API, as decisions about
+scaling nodes up and down, must not be performed based on performance usage at a
+single point in time. Instead the operator will have to gather and store usage
+data for a certain period of time, to decide if scaling is required.
+
+As an example with a configured `timeWindow` (see
+[configuration](#control-plane-machineset-operator-configuration)) of 30 minutes
+for scaling up, the operator has to keep querying the metrics API and averaging
+values until enough data has been aggregated to know the load average for the
+last 30 minutes.
+
+If the metrics API returns it's own window of ~60 second aggregations, the
+operators will have to retrieve 30 values (one per minute), before any scaling
+decision should be made.
+
+#### Hypershift
+
+Hypershift does not use `ControlPlaneMachineSet` to manage the control-planes
+for hosted clusters, as those are not running on their own virtual machines.
+
+Because of this, no special handling should be required.
+
+### Risks and Mitigations
+
+#### Risk 1: Higher costs when scaling up
+
+Scaling up will increase costs for running the cluster, as bigger nodes will
+incur higher costs from the cloud provider.
```

Comment:
probably worth copying down "and mitigations" here, like explaining how `instancesSizes` sets a cap on the largest acceptable size.  And possibly adding alerts around "you've pegged this autoscaler" (...and we're currently ok with that, but you may not be?  ...and we currently wish we had a higher size available, because we're swamped!  Other subclasses?).